src/core/Ekubo.sol

```diff
@@ -0,0 +1,459 @@
+        uint256 hashMod,
+        bytes memory fills,
+        ISignatureTransfer.PermitTransferFrom memory permit,
+        bytes memory sig,
+        uint256 amountOutMin
+    ) internal returns (uint256 buyAmount) {
+        bytes memory data = Encoder.encodeVIP(
+            uint32(IEkuboCore.lock.selector),
+            recipient,
+            feeOnTransfer,
+            hashMul,
+            hashMod,
+            fills,
+            permit,
+            sig,
+            _isForwarded(),
+            amountOutMin
+        );
+        bytes memory encodedBuyAmount =
+            _setOperatorAndCall(address(CORE), data, uint32(IEkuboCallbacks.locked.selector), _ekuboLockCallback);
+        // buyAmount = abi.decode(abi.decode(encodedBuyAmount, (bytes)), (uint256));
+        assembly ("memory-safe") {
+            // We can skip all the checks performed by `abi.decode` because we know that this is the
+            // verbatim result from `locked` and that `locked` encoded the buy amount
+            // correctly.
+            buyAmount := mload(add(0x60, encodedBuyAmount))
+        }
+    }
+
+    function _ekuboLockCallback(bytes calldata data) private returns (bytes memory) {
+        // We know that our calldata is well-formed. Therefore, the first slot is ekubo lock id,
+        // second slot is 0x20 and third is the length of the strict ABIEncoded payload
+        assembly ("memory-safe") {
+            data.length := calldataload(add(0x40, data.offset))
+            data.offset := add(0x60, data.offset)
+        }
+        return locked(data);
+    }
+
+    function _ekuboPay(
+        IERC20 sellToken,
+        address payer,
+        uint256 sellAmount,
+        ISignatureTransfer.PermitTransferFrom calldata permit,
+        bool isForwarded,
+        bytes calldata sig
+    ) private returns (uint256 payment) {
+        if (sellToken == ETH_ADDRESS) {
+            SafeTransferLib.safeTransferETH(payable(msg.sender), sellAmount);
+            return sellAmount;
+        } else {
+            // Encode the call plus the extra data that is going to be needed in the callback
+            bytes memory data;
+            assembly ("memory-safe") {
+                data := mload(0x40)
+
+                mstore(add(data, 0x20), 0x0c11dedd) // selector for pay(address)
+                mstore(add(data, 0x40), sellToken)
+                mstore(add(data, 0x60), payer)
+                mstore(add(data, 0x80), sellAmount)
+                calldatacopy(add(data, 0xa0), permit.offset, 0x80)
+                mstore(add(data, 0x120), isForwarded)
+                mstore(add(data, 0x140), sig.length)
+                calldatacopy(add(data, 0x160), sig.offset, sig.length)
+
+                let size := add(0x124, sig.length)
+
+                // update data length
+                mstore(data, size)
+
+                // update free memory pointer (word-aligned)
+                mstore(0x40, add(data, and(add(add(size, 32), 31), not(31))))
```

Comment:
We don't care that the free memory pointer is word-aligned. Also combine constants where possible, use hex constants, and put constants on the left of commutative operators